components/eamxx/src/physics/mam/eamxx_mam_aci_process_interface.cpp

```diff
@@ -0,0 +1,1326 @@
+    for(int j = 0; j < 2; ++j) {
+      Kokkos::resize(raercol_cw_[i][j], ncol_, mam4::ndrop::ncnst_tot);
+      Kokkos::resize(raercol_[i][j], ncol_, mam4::ndrop::ncnst_tot);
+    }
+  }
+
+  // nact : fractional aero. number activation rate [/s]
+  Kokkos::resize(nact_, ncol_, nlev_, mam_coupling::num_aero_modes());
+
+  // mact : fractional aero. mass activation rate [/s]
+  Kokkos::resize(mact_, ncol_, nlev_, mam_coupling::num_aero_modes());
+
+  // Eddy diffusivity of heat at the interfaces
+  Kokkos::resize(kvh_int_, ncol_, nlev_ + 1);
+
+  // Vertical velocity variance at the interfaces
+  Kokkos::resize(w_sec_int_, ncol_, nlev_ + 1);
+  // Allocate work arrays
+  for(int icnst = 0; icnst < mam4::ndrop::ncnst_tot; ++icnst) {
+    qqcw_fld_work_[icnst] = view_2d("qqcw_fld_work_", ncol_, nlev_);
+  }
+  state_q_work_ =
+      view_3d("state_q_work_", ncol_, nlev_, mam4::aero_model::pcnst);
+
+  //---------------------------------------------------------------------------------
+  // Diagnotics variables from the hetrozenous ice nucleation scheme
+  //---------------------------------------------------------------------------------
+
+  for(int i = 0; i < hetro_scratch_; ++i)
+    Kokkos::resize(diagnostic_scratch_[i], ncol_, nlev_);
+
+  //---------------------------------------------------------------------------------
+  // Initialize the processes
+  //---------------------------------------------------------------------------------
+
+  mam4::AeroConfig aero_config;
+  // configure the nucleation parameterization
+  mam4::NucleateIce::Config nucleate_ice_config;
+  nucleate_ice_.init(aero_config, nucleate_ice_config);
+
+  // configure the heterogeneous freezing parameterization
+  mam4::Hetfrz::Config hetfrz_config;
+  hetfrz_.init(aero_config, hetfrz_config);
+
+  //---------------------------------------------------------------------------------
+  // Setup preprocessing and post processing
+  //---------------------------------------------------------------------------------
+  // set up our preprocess  and postprocess functors
+  preprocess_.initialize(ncol_, nlev_, wet_atm_, wet_aero_, dry_atm_,
+                         dry_aero_);
+
+  postprocess_.initialize(ncol_, nlev_, wet_atm_, wet_aero_, dry_atm_,
+                          dry_aero_);
+
+}  // end function initialize_impl
+
+// ================================================================
+//  RUN_IMPL
+// ================================================================
+void MAMAci::run_impl(const double dt) {
+  const auto scan_policy = ekat::ExeSpaceUtils<
+      KT::ExeSpace>::get_thread_range_parallel_scan_team_policy(ncol_, nlev_);
+
+  // preprocess input -- needs a scan for the calculation of local derivied
+  // quantities
+  Kokkos::parallel_for("preprocess", scan_policy, preprocess_);
+  Kokkos::fence();
+
+  haero::ThreadTeamPolicy team_policy(ncol_, Kokkos::AUTO);
+
+  compute_w0_and_rho(team_policy, dry_atm_, top_lev_, nlev_,
```

Comment:
I'm curious: why are you dispatching several kernels rather than a single monolithic one? You may be able to use smaller temporaries if you have a single kernel instead of many.

Reply:
I talked with @overfelt and he suggested that we may need to revisit this when we optimize the code. If ACI is not taking too much memory, we can have these small kernels launched simultaneously (i.e., without Kokkos::fence) and improve performance. If we are limited by memory, we might combine these kernels.